### DISCUSSION TITLE
WinUI 3 UI testing library

### DISCUSSION THREAD

**Discussion by LeftTwixWand** [microsoft/microsoft-ui-xaml#7655] [Ideas]
Hi. I really like WinUI 3 idea, but its infrastructure is still poor.
I mean, the UI XAML editor in VS 2022, which doesn't work with WinUI 3 markup.
And the same with UI testing frameworks. It's really problematic to create and set-up some UI tests for the WinUI 3 app.
Now, one of the ways to do it is [WinAppDriver](https://github.com/microsoft/winappdriver), but it's not really the tool, that's comfortable to use.
It's difficult to set up and write some tests. WinAppDriver works only with FrameworkElement abstraction level and doesn't know anything about the elements, we're using, like buttons etc.

So I propose to create a library, based on the WinAppDriver library with abstraction for all WinUI elements end encapsulate some casting inside. 
For example:
```CSharp
var buttonOk = session.GetElement<Button>("MyButtonOk");
buttonOk.Command.Execute(null);
```

**Comment by llothar** [comment]
For the last 20 years as active Desktop developer on all relevant platforms my advise is still: Don't do automatic GUI Testing on your final product. Write special apps for each single widgets and emulate the incoming events with calls and make this apps as convenient for manual testing. When you use MVVM  it is all good enough.

**Reply by LeftTwixWand** [reply]
Ok, maybe I was a bit incorrect in the term of `UI Testing`, I would be better to say Integration Testing.

**Comment by marcelwgn** [comment]
I think the only way so far is to use WinAppDriver which admittedly is not the best to use. I do this in my projects (not so much though given the complexity) however one tool to make this easier is [Legerity](https://made-apps.github.io/legerity/index.html) developed by @jamesmcroft. It provides good abstractions around the controls without you having to manually see how to invoke specific functionality of such controls.

**Comment by jamesmcroft** [comment]
@LeftTwixWand if you want to know more about **[Legerity](https://github.com/MADE-Apps/legerity)** let me know. I built it out of the same thought process as your own. 

I think we're also desperately needing a pretty decent update to the `WinAppDriver` too. Not sure when that might happen mind 🤷🏻

**Reply by LeftTwixWand** [reply]
Thanks, this really useful to see, how the other UI testing frameworks work.

**Reply by LeftTwixWand** [reply]
I've checked it, and it's exactly what I was looking for. The idea with wrappers is great.
I would like to see some official support from the WinUI team for such type of project, or at least some roadmap with their plan for it.
Because testing becomes a huge part of development process, and such technology, like WinUI (which is obviously the future of the Windows development) has to be fully equipped with testing infrastructure.
